### PR TITLE
add mima, version scheme

### PR DESCRIPTION
**Pull request by tpolecat**
This adds two things to the default build configuration, with the goal of making it safer for us to evolve our libraries independently and make it clear when downstream updates are and are not compatible.

- Artifacts we publish will now include metadata about the version scheme we're using, which will help SBT detect cases where evictions are not legal. For example, if you end up depending on lucuma-woozle versions 1.2.3 and 1.2.4, it will just use 1.2.4; but if you're depending o 1.2.3 and 4.5.6, sbt will fail the build because the version scheme says these versions are incompatible. The scheme I chose is early-semver (described [here](https://www.scala-sbt.org/1.x/docs/Publishing.html#Version+scheme)) which is what most Scala libraries use.
- Builds will now include [MiMa](https://github.com/lightbend/mima) which performs binary compatibility checks against previous artifacts, to ensure that the assigned version is legal. It also includes [mima-version-check](https://github.com/ChristopherDavenport/sbt-mima-version-check) which automatically computes the set of prior artifact versions, simply by walking back the current version number. **Any library with a version ≥ 0.1.0 should include `mimaReportBinaryIssues` in CI.**

### CODE REVIEW

File: src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala [hunk 1]
```diff
@@ -63,9 +62,6 @@ object LucumaPlugin extends AutoPlugin {
 
   import autoImport._
 
-  override def requires: Plugins =
```
Comment (cquiroz) [inline src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala:66]: Why was this removed?

**Comment by rpiaggio** [comment]
This is awesome, thank you!

Does this mean we should start giving proper version numbers (ie: >= `1.0.0`) to our eternally `0.X.Y` libraries?

> Any library with a version ≥ 0.1.0 should include mimaReportBinaryIssues in CI.

Can you expand on this?

File: src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala (hunk 1 above)
Comment (tpolecat) [reply src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala:66]: It's apparently only required when you need to specify a particular initialization order, otherwise it's unnecessary. I removed it and there were no obvious issues, but I can put it back in if you think we need it.

**Comment by tpolecat** [comment]
With `early-semver` versions 0.1.1 and 0.1.2 are compatible but 0.1.2 and 0.2.0 are not. This is what most Scala libraries do (think http4s 0.21 vs 0.23). Then at 1.x everything is bincompat until 2.x, like regular semantic versioning. So in this scheme 0.0.x is the only series without any bincompat guarantees.

Which is why I suggest that once a library hits 0.1 we should add bincompat checking, because the declared version scheme is claiming that this is what we're doing.

Does that make sense?

**Comment by tpolecat** [comment]
We're going to roll through a lot of version numbers this way but I think this is fine.

**Comment by rpiaggio** [comment]
> With early-semver versions 0.1.1 and 0.1.2 are compatible but 0.1.2 and 0.2.0 are not

Ah I guess that's what the `early` means. Thank you for the explanation.

**Review by rpiaggio — APPROVED** [review]
LGTM

File: src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala (hunk 1 above)
Comment (cquiroz) [reply src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala:66]: Nah, I was just curious

Comment (tpolecat) [reply src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala:66]: Do you have an opinion on this? When I first wrote the plugin I just copied an example without understanding what was going on. This PR continues that trend by copying from another example that doesn't do it.

Comment (cquiroz) [reply src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala:66]: Not really. I just thought it was required but if not, then better

Comment (swalker2m) [reply src/main/scala/edu/gemini/gsp/sbtplugin/GspPlugin.scala:66]: Hey! _I_ wrote the plugin by copying an example without understanding what was going on!

**Comment by tpolecat** [comment]
Note that we can only check bincompat in CI for **releases**, not for PRs (otherwise we could never merge a PR that breaks bincompat). I'm working on a solution to this.
